File: app/analysis.py
"""Post-session analysis: where are you losing time, and why.

Given a target lap and a reference lap (each a LapTrace of position + speed +
throttle + brake + time), this:
  1. builds cumulative distance for each lap from position
  2. resamples both onto a common distance grid
  3. computes the delta-time trace (target time - reference time vs distance)
  4. segments the lap into loss zones (where you bleed time fastest)
  5. diagnoses each zone: braking point, min corner speed, throttle-on point,
     and coasting — turning "0.3s slower in sector 2" into an actionable note.

numpy is fine here: this is post-session, not the live hot path.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import List, Optional

import numpy as np


@dataclass
class LapTrace:
    """One lap's per-tick channels. x/z used for distance; speed in km/h; pedals 0-100."""
    x: List[float]
    z: List[float]
    speed: List[float]
    throttle: List[float]
    brake: List[float]
    t_ms: List[float]                       # cumulative time from lap start, ms
    label: str = ""

    def arrays(self):
        return (np.asarray(self.x, float), np.asarray(self.z, float),
                np.asarray(self.speed, float), np.asarray(self.throttle, float),
                np.asarray(self.brake, float), np.asarray(self.t_ms, float))


@dataclass
class ImprovementZone:
    start_m: float
    end_m: float
    time_lost_s: float
    sector: Optional[int]
    notes: List[str] = field(default_factory=list)

    def to_dict(self) -> dict:
        return {"start_m": round(self.start_m, 0), "end_m": round(self.end_m, 0),
                "time_lost_s": round(self.time_lost_s, 3), "sector": self.sector,
                "notes": self.notes}


def _cum_dist(x, z):
    d = np.zeros(len(x))
    if len(x) > 1:
        seg = np.hypot(np.diff(x), np.diff(z))
        d[1:] = np.cumsum(seg)
    return d


def _resample(dist, channel, grid):
    # ensure strictly increasing distance for interp
    keep = np.concatenate(([True], np.diff(dist) > 0))
    return np.interp(grid, dist[keep], channel[keep])


def _sector_of(d, sector_bounds):
    if not sector_bounds:
        return None
    for i, b in enumerate(sector_bounds):
        if d <= b:
            return i + 1
    return len(sector_bounds) + 1


def analyze(target: LapTrace, reference: LapTrace,
            sector_bounds_m: Optional[List[float]] = None,
            n_grid: int = 800, n_zones: int = 5,
            min_zone_loss_s: float = 0.03) -> dict:
    tx, tz, ts, tth, tbr, tt = target.arrays()
    rx, rz, rs, rth, rbr, rt = reference.arrays()

    td, rd = _cum_dist(tx, tz), _cum_dist(rx, rz)
    L = min(td[-1], rd[-1])
    grid = np.linspace(0, L, n_grid)

    # time-at-distance, then delta
    t_time = _resample(td, tt, grid)
    r_time = _resample(rd, rt, grid)
    delta = (t_time - r_time) / 1000.0          # seconds; positive = target slower
    delta -= delta[0]                            # zero at start line
    total_delta = float(delta[-1])

    # channels on the common grid
    t_speed = _resample(td, ts, grid)
    r_speed = _resample(rd, rs, grid)
    t_brake = _resample(td, tbr, grid)
    r_brake = _resample(rd, rbr, grid)
    t_thr = _resample(td, tth, grid)
    r_thr = _resample(rd, rth, grid)

    # rate of time loss; positive => losing here
    dloss = np.diff(delta, prepend=delta[0])

    # group contiguous losing regions
    losing = dloss > 0
    zones: List[ImprovementZone] = []
    i = 0
    while i < len(grid):
        if not losing[i]:
            i += 1
            continue
        j = i
        while j < len(grid) and losing[j]:
            j += 1
        lost = float(delta[j - 1] - delta[i])
        if lost >= min_zone_loss_s:
            zones.append(_diagnose(grid, i, j, lost, sector_bounds_m,
                                   t_speed, r_speed, t_brake, r_brake, t_thr, r_thr))
        i = j

    zones.sort(key=lambda z: z.time_lost_s, reverse=True)
    top = zones[:n_zones]

    return {
        "target": target.label, "reference": reference.label,
        "total_delta_s": round(total_delta, 3),
        "lap_length_m": round(float(L), 0),
        "n_loss_zones": len(zones),
        "improvements": [z.to_dict() for z in top],
        "delta_trace": {"dist_m": [round(float(d), 1) for d in grid[::20]],
                        "delta_s": [round(float(v), 3) for v in delta[::20]]},
    }


def _diagnose(grid, i, j, lost, sectors,
              t_speed, r_speed, t_brake, r_brake, t_thr, r_thr) -> ImprovementZone:
    seg = slice(i, j)
    start_m, end_m = float(grid[i]), float(grid[j - 1])
    notes: List[str] = []

    BRAKE_TH, THR_TH = 8.0, 80.0

    def first_dist(mask):
        idx = np.nonzero(mask[seg])[0]
        return grid[i + idx[0]] if len(idx) else None

    # braking point: where each first gets meaningfully on the brakes
    tb = first_dist(t_brake > BRAKE_TH)
    rb = first_dist(r_brake > BRAKE_TH)
    if tb is not None and rb is not None:
        d = tb - rb
        if d < -4:
            notes.append(f"Braking {abs(d):.0f} m too early")
        elif d > 4:
            notes.append(f"Braking {d:.0f} m later than reference (good — or overdriving)")

    # min corner speed
    t_min = float(np.min(t_speed[seg]))
    r_min = float(np.min(r_speed[seg]))
    if r_min - t_min > 3:
        notes.append(f"Carrying {r_min - t_min:.0f} km/h less mid-corner")

    # throttle-on point (first time near-full throttle)
    tt_on = first_dist(t_thr > THR_TH)
    rt_on = first_dist(r_thr > THR_TH)
    if tt_on is not None and rt_on is not None and tt_on - rt_on > 5:
        notes.append(f"Back to full throttle {tt_on - rt_on:.0f} m later")

    # coasting (off both pedals)
    coast = np.mean((t_thr[seg] < 5) & (t_brake[seg] < 5))
    if coast > 0.18:
        notes.append(f"Coasting {coast*100:.0f}% of this zone — commit earlier")

    if not notes:
        notes.append("Small line/exit loss")

    return ImprovementZone(start_m=start_m, end_m=end_m, time_lost_s=lost,
                           sector=_sector_of(start_m, sectors), notes=notes)


def format_report(rep: dict) -> str:
    lines = [f"Post-session: {rep['target']} vs {rep['reference']}  "
             f"({rep['lap_length_m']:.0f} m lap)",
             f"Total delta: {rep['total_delta_s']:+.3f} s   "
             f"({rep['n_loss_zones']} loss zones)\n",
             "Top areas to find time:"]
    for n, z in enumerate(rep["improvements"], 1):
        sect = f" [S{z['sector']}]" if z["sector"] else ""
        lines.append(f"  {n}. {z['start_m']:.0f}-{z['end_m']:.0f} m{sect}  "
                     f"−{z['time_lost_s']:.3f} s")
        for note in z["notes"]:
            lines.append(f"       • {note}")
    return "\n".join(lines)

File: flutter/lib/core/analysis.dart
/// Post-session analysis — the parity-locked Dart twin of `app/analysis.py`.
///
/// Given a target lap and a reference lap (each a [LapTrace] of position +
/// speed + pedals + time), this builds cumulative distance, resamples both onto
/// a common distance grid, computes the delta-time trace, segments the lap into
/// loss zones, and diagnoses each zone. [comparisonTraces] returns the per-
/// channel arrays + the racing line (x/z) on the same grid so a single cursor
/// distance links every chart and the map.
///
/// The numeric outputs are rounded to the same decimals as the Python
/// reference so they compare clean (within the harness's float tolerance)
/// against the generated vectors. The resampler matches `numpy.interp`
/// (clamped ends, linear between strictly-increasing samples).
library;

import 'dart:math';

class LapTrace {
  final List<double> x;
  final List<double> z;
  final List<double> speed; // km/h
  final List<double> throttle; // 0-100
  final List<double> brake; // 0-100
  final List<double> tMs; // cumulative ms from lap start
  final String label;

  const LapTrace({
    required this.x,
    required this.z,
    required this.speed,
    required this.throttle,
    required this.brake,
    required this.tMs,
    this.label = '',
  });

  factory LapTrace.fromJson(Map<String, dynamic> j) {
    List<double> a(String k) =>
        ((j[k] as List?) ?? const []).map((e) => (e as num).toDouble()).toList();
    return LapTrace(
      x: a('x'),
      z: a('z'),
      speed: a('speed'),
      throttle: a('throttle'),
      brake: a('brake'),
      tMs: a('t_ms'),
      label: j['label'] as String? ?? '',
    );
  }

  bool get usable => x.length >= 3;
}

// --------------------------------------------------------------- primitives

double _round(double v, int nd) {
  if (!v.isFinite) return v;
  final p = pow(10, nd).toDouble();
  return (v * p).round() / p;
}

List<double> _cumDist(List<double> x, List<double> z) {
  final d = List<double>.filled(x.length, 0.0);
  for (var i = 1; i < x.length; i++) {
    final dx = x[i] - x[i - 1], dz = z[i] - z[i - 1];
    d[i] = d[i - 1] + sqrt(dx * dx + dz * dz);
  }
  return d;
}

List<double> _linspace(double a, double b, int n) {
  if (n <= 1) return [a];
  final out = List<double>.filled(n, 0.0);
  final step = (b - a) / (n - 1);
  for (var i = 0; i < n; i++) {
    out[i] = a + step * i;
  }
  out[n - 1] = b; // exact endpoint, like numpy.linspace(endpoint=True)
  return out;
}

/// numpy.interp-equivalent for a strictly-increasing-filtered series.
List<double> _resample(
    List<double> dist, List<double> chan, List<double> grid) {
  // keep index 0, then any point whose original distance increased
  final xs = <double>[];
  final ys = <double>[];
  for (var i = 0; i < dist.length; i++) {
    if (i == 0 || dist[i] > dist[i - 1]) {
      xs.add(dist[i]);
      ys.add(chan[i]);
    }
  }
  final out = List<double>.filled(grid.length, 0.0);
  if (xs.isEmpty) return out;
  for (var g = 0; g < grid.length; g++) {
    final xq = grid[g];
    if (xq <= xs.first) {
      out[g] = ys.first;
    } else if (xq >= xs.last) {
      out[g] = ys.last;
    } else {
      var lo = 0, hi = xs.length - 1;
      while (hi - lo > 1) {
        final mid = (lo + hi) >> 1;
        if (xs[mid] <= xq) {
          lo = mid;
        } else {
          hi = mid;
        }
      }
      final t = (xq - xs[lo]) / (xs[hi] - xs[lo]);
      out[g] = ys[lo] + t * (ys[hi] - ys[lo]);
    }
  }
  return out;
}

int? _sectorOf(double d, List<double>? sectors) {
  if (sectors == null || sectors.isEmpty) return null;
  for (var i = 0; i < sectors.length; i++) {
    if (d <= sectors[i]) return i + 1;
  }
  return sectors.length + 1;
}

double _minOf(List<double> a, int i, int j) {
  var m = a[i];
  for (var k = i + 1; k < j; k++) {
    if (a[k] < m) m = a[k];
  }
  return m;
}

// ------------------------------------------------------------------ analyze

Map<String, dynamic> analyze(
  LapTrace target,
  LapTrace reference, {
  List<double>? sectorBoundsM,
  int nGrid = 800,
  int nZones = 5,
  double minZoneLossS = 0.03,
}) {
  final td = _cumDist(target.x, target.z);
  final rd = _cumDist(reference.x, reference.z);
  final L = min(td.last, rd.last);
  final grid = _linspace(0, L, nGrid);

  final tTime = _resample(td, target.tMs, grid);
  final rTime = _resample(rd, reference.tMs, grid);
  final delta = List<double>.generate(
      grid.length, (i) => (tTime[i] - rTime[i]) / 1000.0);
  final d0 = delta[0];
  for (var i = 0; i < delta.length; i++) {
    delta[i] -= d0;
  }
  final totalDelta = delta.last;

  final tSpeed = _resample(td, target.speed, grid);
  final rSpeed = _resample(rd, reference.speed, grid);
  final tBrake = _resample(td, target.brake, grid);
  final rBrake = _resample(rd, reference.brake, grid);
  final tThr = _resample(td, target.throttle, grid);
  final rThr = _resample(rd, reference.throttle, grid);

  // rate of time loss; dloss[i] = delta[i] - delta[i-1] (prepend delta[0])
  final zones = <Map<String, dynamic>>[];
  var i = 0;
  while (i < grid.length) {
    final prev = i == 0 ? delta[0] : delta[i - 1];
    if (!(delta[i] - prev > 0)) {
      i++;
      continue;
    }
    var j = i;
    while (j < grid.length) {
      final p = j == 0 ? delta[0] : delta[j - 1];
      if (!(delta[j] - p > 0)) break;
      j++;
    }
    final lost = delta[j - 1] - delta[i];
    if (lost >= minZoneLossS) {
      zones.add(_diagnose(grid, i, j, lost, sectorBoundsM, tSpeed, rSpeed,
          tBrake, rBrake, tThr, rThr));
    }
    i = j;
  }

  zones.sort((a, b) =>
      (b['time_lost_s'] as double).compareTo(a['time_lost_s'] as double));
  final top = zones.take(nZones).toList();

  final dist = <double>[];
  final dlt = <double>[];
  for (var k = 0; k < grid.length; k += 20) {
    dist.add(_round(grid[k], 1));
    dlt.add(_round(delta[k], 3));
  }

  return {
    'target': target.label,
    'reference': reference.label,
    'total_delta_s': _round(totalDelta, 3),
    'lap_length_m': _round(L, 0),
    'n_loss_zones': zones.length,
    'improvements': top,
    'delta_trace': {'dist_m': dist, 'delta_s': dlt},
  };
}

Map<String, dynamic> _diagnose(
    List<double> grid,
    int i,
    int j,
    double lost,
    List<double>? sectors,
    List<double> tSpeed,
    List<double> rSpeed,
    List<double> tBrake,
    List<double> rBrake,
    List<double> tThr,
    List<double> rThr) {
  final startM = grid[i], endM = grid[j - 1];
  final notes = <String>[];
  const brakeTh = 8.0, thrTh = 80.0;

  double? firstDist(bool Function(int) pred) {
    for (var k = i; k < j; k++) {
      if (pred(k)) return grid[k];
    }
    return null;
  }

  final tb = firstDist((k) => tBrake[k] > brakeTh);
  final rb = firstDist((k) => rBrake[k] > brakeTh);
  if (tb != null && rb != null) {
    final d = tb - rb;
    if (d < -4) {
      notes.add('Braking ${d.abs().toStringAsFixed(0)} m too early');
    } else if (d > 4) {
      notes.add(
          'Braking ${d.toStringAsFixed(0)} m later than reference (good — or overdriving)');
    }
  }

  final tMin = _minOf(tSpeed, i, j);
  final rMin = _minOf(rSpeed, i, j);
  if (rMin - tMin > 3) {
    notes.add('Carrying ${(rMin - tMin).toStringAsFixed(0)} km/h less mid-corner');
  }

  final ttOn = firstDist((k) => tThr[k] > thrTh);
  final rtOn = firstDist((k) => rThr[k] > thrTh);
  if (ttOn != null && rtOn != null && ttOn - rtOn > 5) {
    notes.add('Back to full throttle ${(ttOn - rtOn).toStringAsFixed(0)} m later');
  }

  var coastN = 0, coastTot = 0;
  for (var k = i; k < j; k++) {
    coastTot++;
    if (tThr[k] < 5 && tBrake[k] < 5) coastN++;
  }
  final coast = coastTot == 0 ? 0.0 : coastN / coastTot;
  if (coast > 0.18) {
    notes.add('Coasting ${(coast * 100).toStringAsFixed(0)}% of this zone — commit earlier');
  }

  if (notes.isEmpty) notes.add('Small line/exit loss');

  return {
    'start_m': _round(startM, 0),
    'end_m': _round(endM, 0),
    'time_lost_s': _round(lost, 3),
    'sector': _sectorOf(startM, sectors),
    'notes': notes,
  };
}

// -------------------------------------------------------- comparison traces

Map<String, dynamic> comparisonTraces(
  LapTrace target,
  LapTrace reference, {
  int nGrid = 400,
  int nOut = 200,
}) {
  final td = _cumDist(target.x, target.z);
  final rd = _cumDist(reference.x, reference.z);
  final L = min(td.last, rd.last);
  final grid = _linspace(0, L, nGrid);

  final tTime = _resample(td, target.tMs, grid);
  final rTime = _resample(rd, reference.tMs, grid);
  final delta = List<double>.generate(
      grid.length, (i) => (tTime[i] - rTime[i]) / 1000.0);
  final d0 = delta[0];
  for (var i = 0; i < delta.length; i++) {
    delta[i] -= d0;
  }

  final tSpeed = _resample(td, target.speed, grid);
  final rSpeed = _resample(rd, reference.speed, grid);
  final tThr = _resample(td, target.throttle, grid);
  final rThr = _resample(rd, reference.throttle, grid);
  final tBrk = _resample(td, target.brake, grid);
  final rBrk = _resample(rd, reference.brake, grid);
  final lineX = _resample(td, target.x, grid);
  final lineZ = _resample(td, target.z, grid);

  final step = max(1, nGrid ~/ nOut);
  List<double> r(List<double> a, int nd) {
    final out = <double>[];
    for (var k = 0; k < a.length; k += step) {
      out.add(_round(a[k], nd));
    }
    return out;
  }

  return {
    'available': true,
    'target': target.label,
    'reference': reference.label,
    'lap_length_m': _round(L, 0),
    'dist_m': r(grid, 1),
    't_speed': r(tSpeed, 1),
    'r_speed': r(rSpeed, 1),
    't_throttle': r(tThr, 0),
    'r_throttle': r(rThr, 0),
    't_brake': r(tBrk, 0),
    'r_brake': r(rBrk, 0),
    'delta_s': r(delta, 3),
    'line_x': r(lineX, 1),
    'line_z': r(lineZ, 1),
  };
}